pravega/pravega-operator
# Issue 91: Pravega metadata cleanup

## Change log description
- Leverage the k8s finalizer to notify Pravega operator before Pravega object is deleted
- Run cleanup routine right after all Pravega cluster pods are terminated
- Make tests run regardless of the order of execution

## Purpose of the change

Fix #91 
Fix #9
Fix #120 

## How to test the change

Checkout to source branch, follow the instructions in the README.md to deploy a Pravega Cluster.

Use port forward to forward a zookeeper node to localhost for testing

```

$kubectl port-forward [zk-node] 2181:2181

```

Install [zk-shell](https://github.com/rgs1/zk_shell) to log into zookeeper to see the changes.

You can use `ls` or `tree` command in zk-shell to examine znodes.

After Pravega Cluster is created, there are two znodes in the root.

```
$zk-shell localhost 2181
(CONNECTED [localhost,2181]) /> ls
pravega
zookeeper
```
The /pravega is used by Pravega Cluster, the /zookeeper is used by zookeeper itself. 

Delete the Pravega Cluster
```
$kubectl delete -f pravega.yaml
```
Check the znode again
```
(CONNECTED [localhost,2181]) /> ls
zookeeper
```
We can see that the /pravega is deleted after Pravega Cluster resource is deleted.

Signed-off-by: wenqi Wenqi.Mou@emc.com
